metadata/filter.go

```diff
@@ -0,0 +1,110 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metadata
+
+import "strings"
+
+type MetaMap = map[string]string
+
+type metadata struct{}
+
+// Metadata exposes the full implementation
+type Metadata interface {
+	MetaGet
+	MetaCopy
+}
+
+// MetaGet exposes a subset, only Get* methods.
+type MetaGet interface {
+	GetAnnotationsWithPrefix(src MetaMap, prefix string) MetaMap
+	GetLabelsWithPrefix(src MetaMap, prefix string) MetaMap
+}
+
+// MetaCopy exposes a subset, only Copy* methods.
+type MetaCopy interface {
+	CopyAnnotationsWithPrefix(src, dst MetaMap, prefix string)
+	CopyLabelsWithPrefix(src, dst MetaMap, prefix string)
+}
+
+// NewMetadata is a factory function that return a Metadata interface.
+func NewMetadata() Metadata {
+	return metadata{}
+}
+
+// NewMetaGet is a factory function that return a MetaGet interface.
+func NewMetaGet() MetaGet {
+	return metadata{}
+}
+
+// NewMetaCopy is a factory function that return a MetaCopy interface.
+func NewMetaCopy() MetaCopy {
+	return metadata{}
+}
+
+func (metadata) GetAnnotationsWithPrefix(src MetaMap, prefix string) MetaMap {
+	return getMetaMapWithPrefix(src, prefix)
+}
+func (metadata) GetLabelsWithPrefix(src MetaMap, prefix string) MetaMap {
+	return getMetaMapWithPrefix(src, prefix)
+}
+
+func (metadata) CopyAnnotationsWithPrefix(src, dst MetaMap, prefix string) {
+	copyMetaMapWithPrefix(src, dst, prefix)
+}
+func (metadata) CopyLabelsWithPrefix(src, dst MetaMap, prefix string) {
+	copyMetaMapWithPrefix(src, dst, prefix)
+}
+
+// safeCopy copies a key/value pair to dst.
+// When a key is already present in dst, no copy happens.
+func safeCopy(dst MetaMap, key, val string) {
+	if _, err := dst[key]; !err {
+		dst[key] = val
+	}
+}
+
+// copyMetaMapWithPrefix copies key/value pairs in src adding them to dst.
+// When the prefix is not empty, and src key does not contain the prefix string, no copy happens.
+// When a key in src is already present in dst, no copy happens.
+func copyMetaMapWithPrefix(src, dst MetaMap, prefix string) {
+	if prefix != "" {
+		src := getMetaMapWithPrefix(src, prefix)
+		for key, val := range src {
+			safeCopy(dst, key, val)
+		}
+	} else {
+		for key, val := range src {
+			safeCopy(dst, key, val)
+		}
+	}
+}
+
+// getMetaMapWithPrefix returns a map of key/value pairs contained in src that matches the prefix.
+// When the prefix is not empty, and src key does not contain the prefix string, no copy happens.
+func getMetaMapWithPrefix(src MetaMap, prefix string) MetaMap {
+	dst := MetaMap{}
+	if prefix != "" {
```

Comment:
Not really needed as any string starts with `""`, which is fine. Remove this check please.

Reply:
Yeah I know, right?  :sunglasses: 
I hate that condition too... and I'd love to get rid of it, but...
_That condition is 100% required for the code to work as intended_.
It's actually a per performance optimization as much as a sanity check.

Instead of explaining in a comment, I'll provide a quick and dirty implementation in the playground to illustrate why this happens. https://go.dev/play/p/F-k_eRQbPwY

Reply:
```
That condition is 100% required for the code to work as intended.
```
Could you comment on this? Removing the check would have the same behaviour. It would just be slower, but we don't care. We are just filtering around 10 annotations/labels in the worst case. Not a big deal.

If you really want to keep it, rewrite it to something like this:
```
	if prefix == "" {
		return src
	}
	
	for key, val := range src {
		if strings.HasPrefix(key, prefix) {
			dst[key] = val
		}
	}

	return dst
```
Also, AFAIK provided you delete the key you are currently looping over, it's safe to do so in Go. So the function would become:
```
func filterByPrefix(entries map[string]string, prefix string) map[string]string {
	if prefix != "" {
		for key, val := range entries {
			if !strings.HasPrefix(key, prefix) {
				delete(entries, key)
			}
		}
	}
	
	return entries
}
```
I renamed `src` to `entries` as in this version it makes more sense IMHO.

Reply:
Okay, yeah... I refactored it to do the zero-copy NOOP near the top, like the first example you gave.

Reply:
```
func filterByPrefix[M map[K]V, K, V, S ~string](entries M, prefix S) M {
```
I don't understand why to overengineer something that is specifically made for annotations and labels. Those are **always** going to be a `map[string]string`. This makes the code more complex to understand which is what I've been saying on my comments.